Extend AudioSampleBufferCompressor capabilities to allow for any CMSampleBuffer conversion

jyavenard opened the pull request:
#### 10ea3bdfd9009b070aa95cb6e46e7948f0b1b7f6
<pre>
Extend AudioSampleBufferCompressor capabilities to allow for any CMSampleBuffer conversion
<a href="https://bugs.webkit.org/show_bug.cgi?id=283905">https://bugs.webkit.org/show_bug.cgi?id=283905</a>
<a href="https://rdar.apple.com/140782536">rdar://140782536</a>

Reviewed by Youenn Fablet.

We extend the class to be able to decode various format into PCM as required by the WebCodec&apos;s AudioDecoderCocoa class
(WebCodec&apos;s AudioEncoderCocoa).
We also rewrite the AudioConverterComplexInputDataProc used by processSample using
the available CoreMedia methods and how we create a CMSampleBuffer. This new version is significantly
smaller and less complex while being able to handle more format (in particular, in can now handle any planar formats and not a maximum of two).

The addSampleBuffer method now returns a GenericPromise which is resolved
once the sampleBuffer has been fulled decoded and all data has been returned
to the callback.
Decoding or Encoding errors are now reported back to the consumer/

We add a new mode where the converter will now longer generate the timestamps but
instead set the timestamp of the input to be the same as the one on the input
(this is an AudioDecoder requirement)

This change will be fully tested in a follow-up commit where we introduce AudioDecoderCocoa.
For now, there&apos;s no impact on its only consumer: MediaRecorderPrivateEncoder
which is covered by existing tests.

Fly-By: change use of `auto` for explicit `int` when reporting error as it causes compilation error when using OSStatus.

* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::AudioFileReader::getPacketDescriptions const): Deleted.
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.h: Extract getPacketDescriptions method from AudioFileReaderCocoa.
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::getPacketDescriptions):
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
(WebCore::AudioSampleBufferCompressor::isPCM const):
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::flushInternal):
(WebCore::AudioSampleBufferCompressor::setGenerateTimestamp):
(WebCore::AudioSampleBufferCompressor::initAudioConverterForSourceFormatDescription):
(WebCore::AudioSampleBufferCompressor::gradualDecoderRefreshCount):
(WebCore::AudioSampleBufferCompressor::sampleBuffer):
(WebCore::AudioSampleBufferCompressor::provideSourceDataNumOutputPackets):
(WebCore::AudioSampleBufferCompressor::setTimeFromSample):
(WebCore::AudioSampleBufferCompressor::processSampleBuffers):
(WebCore::AudioSampleBufferCompressor::addSampleBuffer):

Canonical link: <a href="https://commits.webkit.org/287362@main">https://commits.webkit.org/287362@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a235662c03cdb39258662f7628384b5321b4c54a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79264 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58293 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32636 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83890 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30430 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81397 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67384 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6563 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62012 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19909 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82331 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52071 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72226 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42315 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49415 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26398 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28819 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70529 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26832 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85281 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6562 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4559 "Found 1 new test failure: fast/css/value-list-out-of-bounds-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70253 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6725 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68096 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69501 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17332 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13551 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12400 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6519 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12283 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6441 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9908 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8233 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
```diff
@@ -44,10 +44,11 @@ class AudioSampleBufferCompressor : public ThreadSafeRefCountedAndCanMakeThreadS
 
     bool isEmpty() const;
     void setBitsPerSecond(unsigned);
+    void setGenerateTimestamp(bool);
```
Do we have to update this dynamically? How about using a `const bool` instead?
Also, it is probably set in a thread but used in another.

Maybe we could start adding threading ASSERTs and annotations as well in this class.

jyavenard replied:
no it is set on creation ; I can place it in the constructor as an optional.

jyavenard replied:
or make the bool an atomic, it only affects the next conversion.

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -569,17 +515,18 @@
-            return;
+            return GenericPromise::createAndReject();
         protectedThis->processSampleBuffer(buffer.get());
+        return std::exchange(protectedThis->m_lastError, noErr) ? GenericPromise::createAndReject() : GenericPromise::createAndResolve();
```
When there is an error, this is usually terminal for WebCodecs, and there is a need to reconfigure the encoder, which would then create a new internal encoder.
Do we need to std::exchange?

jyavenard replied:
it prevents having to deal with what to do when the converter is in an error state.

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -141,6 +144,8 @@
                 return GenericPromise::createAndReject();
             }
         }
+        m_lastError = noErr;
```
m_lastError is probably only used in `m_serialDispatchQueue`, can we add a threading annotation?

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -307,54 +318,39 @@
 }
 
-RetainPtr<CMSampleBufferRef> AudioSampleBufferCompressor::sampleBuffer(AudioBufferList fillBufferList)
+Expected<RetainPtr<CMSampleBufferRef>, OSStatus> AudioSampleBufferCompressor::sampleBuffer(AudioBufferList* fillBufferList, uint32_t numSamples)
```
Why do we have to pass fillBufferList as an in/out parameter?
Can we simplify and just pass a const WebAudioBufferList&?

jyavenard replied:
We don't, but this is your code so I left that part as is :)

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -373,124 +369,47 @@
+void AudioSampleBufferCompressor::setTimeFromSample(CMSampleBufferRef sample)
+{
+    m_currentNativePresentationTimeStamp = PAL::CMSampleBufferGetPresentationTimeStamp(sample);
+    m_currentOutputPresentationTimeStamp = PAL::CMSampleBufferGetOutputPresentationTimeStamp(sample);
```
Only called in the serial queue as well?

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
```diff
@@ -64,40 +70,43 @@ class AudioSampleBufferCompressor : public ThreadSafeRefCountedAndCanMakeThreadS
-    bool m_isDraining { false };
+    RetainPtr<CMBufferQueueRef> m_outputBufferQueue; // initialized on the caller's thread once, never modified after that.
+    RetainPtr<CMBufferQueueRef> m_inputBufferQueue; // initialized on the caller's thread once, never modified after that.
+    bool m_isEncoding WTF_GUARDED_BY_CAPABILITY(queue().get()) { false };
```
Can we set it to true here and not in constructor?

jyavenard replied:
I'll take this in the rename PR, I missed this comment

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -280,81 +289,71 @@
 }
 
-RetainPtr<CMSampleBufferRef> AudioSampleBufferCompressor::sampleBuffer(AudioBufferList fillBufferList)
+Expected<RetainPtr<CMSampleBufferRef>, int> AudioSampleBufferCompressor::sampleBuffer(const WebAudioBufferList& fillBufferList, uint32_t numSamples)
```
Can we keep OStatus here, this is much easier to read?

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -280,81 +289,71 @@
-            if (auto error = PAL::AudioConverterGetProperty(m_converter, kAudioConverterCompressionMagicCookie, &cookieSize, cookie.data())) {
+            if (int error = PAL::AudioConverterGetProperty(m_converter, kAudioConverterCompressionMagicCookie, &cookieSize, cookie.data()))
                 RELEASE_LOG_ERROR(MediaStream, "AudioSampleBufferCompressor getting kAudioConverterCompressionMagicCookie failed with %d", error);
-                return nil;
```
Why not returning an error anymore?

jyavenard replied:
It's not a fatal error and it will always error on PCM content.

If we do fail to set it, then decoding will error for the codec type that do need it (vorbis, Flac)

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -280,81 +289,71 @@
     }
-
     auto sampleBuffer = adoptCF(rawSampleBuffer);
+    if (int error = PAL::CMSampleBufferSetDataBufferFromAudioBufferList(sampleBuffer.get(), kCFAllocatorDefault, kCFAllocatorDefault, kCMSampleBufferFlag_AudioBufferList_Assure16ByteAlignment, fillBufferList.list())) {
```
I wonder if we have elsewhere this WebAudioBufferList -> CMSampleBuffer conversion. Maybe it could be moved to a utility routine somewhere if that is the case.

jyavenard replied:
I was thinking the same thing. I'll do that in a follow-up. It's a very common way to go about it ever since I discovered `CMSampleBufferSetDataBufferFromAudioBufferList` :)

jyavenard committed:
Extend AudioSampleBufferCompressor capabilities to allow for any CMSampleBuffer conversion
https://bugs.webkit.org/show_bug.cgi?id=283905
rdar://140782536

Reviewed by Youenn Fablet.

We extend the class to be able to decode various format into PCM as required by the WebCodec's AudioDecoderCocoa class
(WebCodec's AudioEncoderCocoa).
We also rewrite the AudioConverterComplexInputDataProc used by processSample using
the available CoreMedia methods and how we create a CMSampleBuffer. This new version is significantly
smaller and less complex while being able to handle more format (in particular, in can now handle any planar formats and not a maximum of two).

The addSampleBuffer method now returns a GenericPromise which is resolved
once the sampleBuffer has been fulled decoded and all data has been returned
to the callback.
Decoding or Encoding errors are now reported back to the consumer/

We add a new mode where the converter will now longer generate the timestamps but
instead set the timestamp of the input to be the same as the one on the input
(this is an AudioDecoder requirement)

This change will be fully tested in a follow-up commit where we introduce AudioDecoderCocoa.
For now, there's no impact on its only consumer: MediaRecorderPrivateEncoder
which is covered by existing tests.

Fly-By: change use of `auto` for explicit `int` when reporting error as it causes compilation error when using OSStatus.

* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::AudioFileReader::getPacketDescriptions const): Deleted.
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.h: Extract getPacketDescriptions method from AudioFileReaderCocoa.
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::getPacketDescriptions):
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
(WebCore::AudioSampleBufferCompressor::isPCM const):
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::flushInternal):
(WebCore::AudioSampleBufferCompressor::setGenerateTimestamp):
(WebCore::AudioSampleBufferCompressor::initAudioConverterForSourceFormatDescription):
(WebCore::AudioSampleBufferCompressor::gradualDecoderRefreshCount):
(WebCore::AudioSampleBufferCompressor::sampleBuffer):
(WebCore::AudioSampleBufferCompressor::provideSourceDataNumOutputPackets):
(WebCore::AudioSampleBufferCompressor::setTimeFromSample):
(WebCore::AudioSampleBufferCompressor::processSampleBuffers):
(WebCore::AudioSampleBufferCompressor::addSampleBuffer):

Canonical link: https://commits.webkit.org/287362@main
